File: services/java_gc/src/main/java/com/backpressure/javagc/MetricsTracker.java
package com.backpressure.javagc;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicIntegerArray;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Thread-safe sliding-window RPS tracker using a circular buffer of per-second counts.
 * Mirrors the algorithm from the TypeScript sinkhole MetricsTracker.
 */
public class MetricsTracker {

    private final int windowSize;
    private final AtomicIntegerArray slots;
    private final AtomicInteger currentSlot;
    private final AtomicLong totalRequests;
    private ScheduledExecutorService scheduler;

    /**
     * Creates a new MetricsTracker with the given window size.
     * @param windowSize number of seconds in the sliding window
     */
    public MetricsTracker(int windowSize) {
        this.windowSize = windowSize;
        this.slots = new AtomicIntegerArray(windowSize);
        this.currentSlot = new AtomicInteger(0);
        this.totalRequests = new AtomicLong(0);
    }

    /** Creates a new MetricsTracker with a default 10-second window. */
    public MetricsTracker() {
        this(10);
    }

    /**
     * Starts the per-second slot rotation timer.
     */
    public void start() {
        if (scheduler != null) {
            return;
        }
        scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "metrics-rotator");
            t.setDaemon(true);
            return t;
        });
        scheduler.scheduleAtFixedRate(() -> {
            int next = (currentSlot.get() + 1) % windowSize;
            slots.set(next, 0);
            currentSlot.set(next);
        }, 1, 1, TimeUnit.SECONDS);
    }

    /**
     * Stops the slot rotation timer.
     */
    public void stop() {
        if (scheduler != null) {
            scheduler.shutdownNow();
            scheduler = null;
        }
    }

    /**
     * Records a single incoming request.
     */
    public void recordRequest() {
        slots.incrementAndGet(currentSlot.get());
        totalRequests.incrementAndGet();
    }

    /**
     * Calculates the current requests per second over the sliding window.
     * @return the average RPS across all window slots
     */
    public double getRps() {
        int sum = 0;
        for (int i = 0; i < windowSize; i++) {
            sum += slots.get(i);
        }
        return (double) sum / windowSize;
    }

    /**
     * Returns the total number of requests recorded since creation.
     * @return the monotonic total request count
     */
    public long getTotalRequests() {
        return totalRequests.get();
    }
}

File: services/simulated_pause/src/metrics_tracker.ts
/**
 * Tracks request rates using a circular buffer of per-second counts
 * over a sliding window.
 */
export class MetricsTracker {
  private readonly windowSize: number;
  private readonly slots: number[];
  private currentSlot: number;
  private totalRequests: number;
  private timer: ReturnType<typeof setInterval> | null;

  /**
   * Creates a new MetricsTracker.
   * @param windowSize - Number of seconds in the sliding window.
   */
  constructor(windowSize = 10) {
    this.windowSize = windowSize;
    this.slots = new Array<number>(windowSize).fill(0);
    this.currentSlot = 0;
    this.totalRequests = 0;
    this.timer = null;
  }

  /**
   * Starts the per-second slot rotation timer.
   */
  start(): void {
    if (this.timer) {
      return;
    }
    this.timer = setInterval(() => {
      this.currentSlot = (this.currentSlot + 1) % this.windowSize;
      this.slots[this.currentSlot] = 0;
    }, 1_000);
  }

  /**
   * Stops the slot rotation timer.
   */
  stop(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }

  /**
   * Records a single incoming request.
   */
  recordRequest(): void {
    // currentSlot is always within bounds due to modulo arithmetic
    this.slots[this.currentSlot] = (this.slots[this.currentSlot] ?? 0) + 1;
    this.totalRequests++;
  }

  /**
   * Calculates the current requests per second over the sliding window.
   * @returns The average RPS across all window slots.
   */
  getRps(): number {
    let sum = 0;
    for (const count of this.slots) {
      sum += count;
    }
    return sum / this.windowSize;
  }

  /**
   * Returns the total number of requests recorded since creation.
   * @returns The monotonic total request count.
   */
  getTotalRequests(): number {
    return this.totalRequests;
  }
}
